src/main/scala/is/hail/variant/VariantDataset.scala

```diff
@@ -127,15 +128,57 @@ class VariantDatasetFunctions(private val vds: VariantDataset) extends AnyVal {
       vds
   }
 
-  def exportGen(path: String, precision: Int = 4) {
-    require(vds.wasSplit)
+  def exportBgen(path: String, nBitsPerProb: Int = 8, parallel: Boolean = false) {
+    val rsidQuery: Option[Querier] = vds.vaSignature.getOption("rsid")
+      .filter {
+        case TString => true
+        case t => warn(
+          s"""found `rsid' field, but it was an unexpected type `$t'.  Emitting missing rsID.
+             |  Expected ${ TString }""".stripMargin)
+          false
+      }.map(_ => vds.queryVA("va.rsid")._2)
+
+    val varidQuery: Option[Querier] = vds.vaSignature.getOption("varid")
+      .filter {
+        case TString => true
+        case t => warn(
+          s"""found `varid' field, but it was an unexpected type `$t'.  Emitting missing variant ID.
+             |  Expected ${ TString }""".stripMargin)
+          false
+      }.map(_ => vds.queryVA("va.varid")._2)
+
+    val sampleIds = vds.stringSampleIds.toArray
+    val partitionSizes = vds.hc.sc.runJob(vds.rdd, getIteratorSize _)
+    val nVariants = partitionSizes.sum
+    val nSamples = vds.nSamples
 
-    def writeSampleFile() {
-      //FIXME: should output all relevant sample annotations such as phenotype, gender, ...
-      vds.hc.hadoopConf.writeTable(path + ".sample",
-        "ID_1 ID_2 missing" :: "0 0 0" :: vds.sampleIds.map(s => s"$s $s 0").toList)
-    }
+    val header = BgenWriter.headerBlock(sampleIds, nVariants)
+
+    vds.rdd.mapPartitionsWithIndex { case (i: Int, it: Iterator[(Variant, (Annotation, Iterable[Genotype]))]) =>
+      val bb = new ArrayBuilder[Byte]
+
+      val partHeader = if (parallel) {
+        BgenWriter.emitHeaderBlock(bb, sampleIds, partitionSizes(i))
+        val r = bb.result()
+        bb.clear()
+        r
```

Comment:
It looks like you're trying to save a clear, but (1) the code is shorter if you clear before emit instead of after result, and (2)  you don't save any because this way you unnecessarily clear after the last result call.